### ISSUE TITLE
(Meta-)gradient computation via multiple calls to backward()

### ISSUE THREAD

**Issue by kylehkhsu** [facebookresearch/higher#7]
To my understanding, in PyTorch, repeated calls to backward() result in a running sum of gradients in a Parameter's .grad attribute.

It's common to use multiple tasks in a meta-batch when updating the meta-parameters to reduce meta-gradient variance. This is done in higher's [MAML example](https://github.com/facebookresearch/higher/blob/master/examples/maml-omniglot.py#L162).

Does this mean that the effective learning rate for the meta-optimization scales proportionally with the meta-batch size? Should we account for this by dividing the query loss by the meta-batch size?

**Comment by egrefen** [comment]
> To my understanding, in PyTorch, repeated calls to backward() result in a running sum of gradients in a Parameter's .grad attribute.

That's my understanding as well.

> Does this mean that the effective learning rate for the meta-optimization scales proportionally with the meta-batch size? Should we account for this by dividing the query loss by the meta-batch size?

Yes, that's [what we do](https://github.com/facebookresearch/higher/blob/master/examples/maml-omniglot.py#L165). If you're interpreting the loss as being a monte-carlo estimate of the expected task loss across tasks, you sample `num_tasks`, compute the sum of task losses, and divide by `num_tasks`. The gradient of the MC estimate of the expectation is the MC estimate of the gradient of the expectation, here.

Closing task, but please re-open if this doesn't clarify things.

**Comment by MurtyShikhar** [comment]
Hi Ed, 
I had a similar question so I thought I should comment on this as well. It seems like in the example, you're dividing by `num_tasks` only for the purposes of logging, but the meta-gradient is still the sum of all gradients instead of being an average (MC estimate)?
